### ISSUE TITLE
[Dream Cycle 2026-08-17] darwin-evolution: ADR-249's cost seam was shipped but never wired into evolve() + weight-eft,learn scan

### ISSUE THREAD

**Issue by ruvnet** [ruvnet/metaharness#206]
MetaHarness Dream Cycle, 2026-08-17. Deep surface: `darwin-evolution` (slot 2). Scan: `weight-eft`, `learn`.

## Tonight's Rotation

```
DATE=2026-08-17  DEEP=darwin-evolution  SCAN=weight-eft,learn  SLOT=2
COMMIT=5453c8c990824b54e05f289774e5a8b2cea0a32e  BRANCH=main
```
Build: full monorepo `npm run build` clean (JS + all workspaces; no wasm/NAPI degradation tonight). `LLM_EVAL=available` (`OPENROUTER_API_KEY` present) — used only by research agents; the candidate itself is $0.

## Ledger Check

Ledger on `main` has only 2 rows (2026-08-13, 2026-08-14) despite 4 nights of prior runs. Verified this is **cosmetic, not a pipeline failure**: `git ls-remote --heads origin` shows `dream/2026-08-15-generator-genome` and `dream/2026-08-16-flywheel-promotion` both exist, and PR #200 / PR #205 are real, open, draft, non-stale (mergeable_state clean, 2/1 days old), each with a full "(evaluated)" writeup and their own ledger-row commit sitting on their own unmerged branch. #181 and #188 (2026-08-13/14) are MERGED. No skipped step found on any prior night — direct evidence checked before reporting.

## Learning Signals (STEP 1.1)

No duplicate finding direction across the last 4 nights (security-adversarial → host-adapters → generator-genome → flywheel-promotion, no repeats). 2/4 candidate PRs merged, 2 open-not-stale — an early "review latency" signal, not yet a pattern (both PRs are <3 days old). No `LLM_EVAL=blocked` streak. Prior-gist score below.

## Prior Gist Score (STEP 1.2) — 9/10

Scored the 2026-08-16 gist (`dream/2026-08-16-flywheel-promotion`, PR #205): benchmark evidence grade A/B (+2), 8 competitor rows (+2), 3 specific recommendations (+2), valid self-verified witness (+2), genuinely novel finding — receipt-splicing fix (+1), **1650 words, over the 1500-word cap (0/1)**. Tonight's own gist was trimmed to 1477 words to fix this specific recurring miss.

## Deep Dive Findings — darwin-evolution

Full detail, hypothesis, evaluation receipt, adversarial critique, and competitor table: [`docs/dream-cycle/2026-08-17-gist.md`](https://github.com/ruvnet/metaharness/blob/dream/2026-08-17-darwin-evolution/docs/dream-cycle/2026-08-17-gist.md).

**Finding**: ADR-249 (2026-08-10) added an opt-in `signals.cost` seam to `scorer.ts`'s `scoreVariant` so a caller could feed a deterministic cost figure into `costEfficiency` instead of the pinned `1.0`. It was tested in isolation (`scorer-signals.test.ts`) but the production `evolve()` loop's only call site (`evaluateVariant`) never passed `signals` — the seam was structurally unreachable. A deterministic parsimony signal (`variantBytes(variant.dir)`, total on-disk surface bytes) already existed in the same file, used only for a Pareto tie-break.

**Candidate**: `EvolutionConfig` gains an optional `costBudgetBytes: number`. When set, `evaluateVariant` computes `variantBytes(variant.dir)` and passes `{ cost: { units, budgetUnits: costBudgetBytes } }` into `scoreVariant`. Omitted (the default, and every existing caller), `signals` stays `undefined` — byte-identical to before. `evaluateVariant` is also now exported (was module-private) for direct testability. Diff: `evolve.ts` +18/-2, `types.ts` +11, 1 test file +79 lines (4 new tests). Well under 300 lines.

## Hypothesis

Given `evaluateVariant`, which computes `variantBytes(variant.dir)` for Pareto tie-breaking but feeds no signal into the ADR-249 cost seam, when `EvolutionConfig.costBudgetBytes` is added and threaded through only when set, then an over-budget variant should score strictly lower `costEfficiency` (and, on a crafted near-tie fixture, this alone should flip a promotion decision), while every existing caller stays byte-identical — subject to zero regression, `scorer.ts` untouched, diff under 300 lines. Full text in the gist.

## Evaluation Receipt

Deterministic, not sampled. Before/after repro: source fix stashed (test kept) → the 4 new tests fail exactly as predicted (`evaluateVariant is not a function`); 5 pre-existing tests unaffected. Fix restored → package suite **632 passed, 14 skipped, 0 regressions**; full monorepo build clean; `tsc --noEmit` clean; 2 dependent packages (`create-agent-harness`, `flywheel`) re-run clean (11/11). New tests assert the arithmetic directly: a 200-byte variant vs. a 100-byte budget decays `costEfficiency` to exactly `round6(100/200)=0.5`; on a crafted fixture (parent `finalScore=0.4`, `promotionDelta=0.02`) that decay alone moves `finalScore` from 0.435 (promoted) to 0.385 (not) — no other gate clause moves.

**Adversarial critique** (independent, no authorship context): verdict **ACCEPT**. Independently re-ran all tests, traced the omitted-field path directly rather than trusting comments, re-derived the promotion-flip arithmetic, and confirmed no read-before-write race (all variant-directory writes are `await`ed before `evaluateVariant` runs). Two disclosed, non-blocking caveats: (1) `variantBytes`'s pre-existing whitespace-stripping gameability becomes score-consequential rather than merely tie-break-consequential once a future caller sets `costBudgetBytes` (none does yet); (2) a pre-existing, unrelated `copyVariantDir` merge-not-clear behavior could inflate byte counts if a `workRoot` is reused across runs with colliding ids.

**Reward hack check**: no test/benchmark weakened, no gold data touched, `variantBytes`'s gameability is a pre-existing property (already used for Pareto tie-break) — disclosed as a caveat, not hidden.

**Security review**: no credential/shell/network surface touched, no new dependency. Also statically audited (unrelated to the candidate, $0): darwin-mode's SWE-bench solvers all `git fetch --depth 1` the base commit only, falling back to `--depth 200` *ancestor* history — no future/gold-patch leak path found, unlike the class of grading defect that hit SWE-bench Pro's public graders.

**Darwin**: not run — this is a scorer-wiring/config-plumbing fix to the evolution loop's own code, not a tunable routing/topology/prompt/tool/tier parameter (ADR-071 allowlist), same reasoning as 2026-08-15/16.

**Flywheel**: not applicable — the candidate modifies `@metaharness/darwin`'s own scorer wiring; durable evidence is the deterministic repro above.

## Scan Findings

**weight-eft** (LoRA fine-tuning of cheap models on agent trajectories, cost-axis cascade play): current SFT/DPO split with train/eval disjointness, reward-hack filtering, and context-truncation guards is broadly aligned with 2025-2026 practice (RSFT/RAFT-style filtering, on-policy DPO), but is behind on two points found tonight — it discards whole failed trajectories rather than mining partial-success signal the way "Exploring Expert Failures" (arXiv:2504.11343, grade B) does, and it has no explicit safeguard yet against multi-generation bootstrap distribution shift (arXiv:2512.11867, grade B) if adapter output ever re-enters the training archive. No comparable productized system found elsewhere (grade C, absence-of-evidence).

**learn** (GEPA-style reflective prompt/scaffold optimization, ADR-228): darwin-mode's `src/gepa/` is a faithful minimal port of GEPA (arXiv:2507.19457 — now **ICLR 2026 Oral**, grade A). It skips the paper's crossover/merge step and the broader `optimize_anything` scope (code/config, not just prompts) that both native `dspy.GEPA` and the standalone `gepa-ai/gepa` package now ship (grade B) — a 2026-current tool would likely integrate one of those rather than maintain a parallel reimplementation. Adjacent: DeepMind's AlphaEvolve (arXiv:2506.13131, grade B).

## Competitors Reviewed

LangGraph, AutoGen (Microsoft Agent Framework), CrewAI, OpenAI Agents SDK/Codex, Claude Agent SDK, DSPy/GEPA — plus, going past the fixed list: Darwin Gödel Machine (Sakana/UBC), "Meta-Harness" (Stanford, arXiv:2603.28052) and "Self-Harness" (arXiv:2606.09498), both 2026-dated academic peers that run an archived, benchmark-gated harness-rewrite loop close to darwin-mode's own. Full table with per-system grading in the gist. Synthesis: MetaHarness is not alone in evolving the harness itself anymore, but no found peer combines that with a signed (Ed25519), third-party-replayable accept+reject promotion ledger.

Independently corroborating this repo's own contamination finding from 2026-08-16: OpenAI retired SWE-bench Verified from its own evals in Feb 2026 (59% of failed-test cases found flawed), and METR (Mar 2026) found ~half of SWE-bench-Verified-passing PRs wouldn't actually be merged by real maintainers.

## Gist

No gist-creation tool is available in this session's toolset (matching 2026-08-13 through 08-16) — committed instead: [`docs/dream-cycle/2026-08-17-gist.md`](https://github.com/ruvnet/metaharness/blob/dream/2026-08-17-darwin-evolution/docs/dream-cycle/2026-08-17-gist.md). `GIST=LOCAL`.

## Witness

```
session_commit : 5453c8c990824b54e05f289774e5a8b2cea0a32e
report_sha256  : b5e71a1f847cc37d43cb4c7b41db718d0832cec67aab701a121a0c3527b69693
witness        : 57ed9167c8085b57799a0479d9842baed13f94e782c889b9f6f00f0804f25a72
```
Verifier procedure is in the gist's Witness section and was independently self-verified (round-trip reproduced programmatically) before this issue was filed.

## Recommendation

**evaluated: accepted** (ACCEPT-WITH-CAVEATS) — real, independently-critiqued, fully deterministic and reproducible fix that closes a genuine "ADR shipped, implementation dangling" gap. Two comparable-severity gaps are disclosed as still-open, natural next-night follow-ups (see gist's Recommended Next Steps): wiring a real `costBudgetBytes` caller, and auditing `variantBytes`'s now-score-consequential gameability. A third, cheap ($1-3, <20 live calls) candidate — checking arXiv:2602.16805's "simple baselines beat evolved scaffolds" claim against darwin-mode's own promoted genome — is flagged for a future night, not attempted tonight to keep this diff single-mechanism and $0.
